plugins/modules/aci_vmm_controller.py

```diff
@@ -327,11 +331,23 @@ def main():
             aci_class="vmmCtrlrP",
             aci_rn="ctrlr-{0}".format(name),
             module_object=name,
-            target_filter={"name": "name"},
+            target_filter={"name": name},
         ),
-        child_classes=["vmmRsMgmtEPg", "vmmRsAcc"],
+        child_classes=child_classes,
     )
 
+    # vmmProvP is not allowed to execute a query with rsp-subtree set in the filter_string
+    # due to complicated url construction logic which should be refactored creating a temporary fix inside module
+    # TODO refactor url construction logic if more occurences of rsp-subtree not supported problem appear
```

Comment:
Just a suggestion for refactoring or to do in the future-
Can we do something like this in construct_url function of aci.py of module_utils?
```
def construct_url(
        self,
        root_class,
        subclass_1=None,
        subclass_2=None,
        subclass_3=None,
        subclass_4=None,
        subclass_5=None,
        child_classes=None,
        config_only=True,
    ):
        """
        This method is used to retrieve the appropriate URL path and filter_string to make the request to the APIC.

        :param root_class: The top-level class dictionary containing aci_class, aci_rn, target_filter, and module_object keys.
        :param sublass_1: The second-level class dictionary containing aci_class, aci_rn, target_filter, and module_object keys.
        :param sublass_2: The third-level class dictionary containing aci_class, aci_rn, target_filter, and module_object keys.
        :param sublass_3: The fourth-level class dictionary containing aci_class, aci_rn, target_filter, and module_object keys.
        :param child_classes: The list of child classes that the module supports along with the object.
        :type root_class: dict
        :type subclass_1: dict
        :type subclass_2: dict
        :type subclass_3: dict
        :type subclass_4: dict
        :type subclass_5: dict
        :type child_classes: list
        :return: The path and filter_string needed to build the full URL.
        """
        self.filter_string = ""

        if child_classes is None:
            self.child_classes = set()
        else:
            self.child_classes = set(child_classes)

        if subclass_5 is not None:
            self._construct_url_6(root_class, subclass_1, subclass_2, subclass_3, subclass_4, subclass_5, config_only)
        elif subclass_4 is not None:
            self._construct_url_5(root_class, subclass_1, subclass_2, subclass_3, subclass_4, config_only)
        elif subclass_3 is not None:
            self._construct_url_4(root_class, subclass_1, subclass_2, subclass_3, config_only)
        elif subclass_2 is not None:
            self._construct_url_3(root_class, subclass_1, subclass_2, config_only)
        elif subclass_1 is not None:
            self._construct_url_2(root_class, subclass_1, config_only)
        else:
            self._construct_url_1(root_class, config_only)

        if self.params.get("port") is not None:
            self.url = "{protocol}://{host}:{port}/{path}".format(path=self.path, **self.module.params)
        else:
            self.url = "{protocol}://{host}/{path}".format(path=self.path, **self.module.params)

        if self.child_classes:
          # new code here ->
          subclasses = [subclass_1, subclass_2, subclass_3, subclass_4, subclass_5]
          
          for i in range(len(subclasses) - 1): 
              if subclasses[i] is not None and subclasses[i + 1] is None and subclasses[i].get("aci_class") in 
              EXCLUDE_RSP_SUBTREE and subclasses[i].get(target_filter) is None:
               self.update_qs(
                          {
                              "rsp-subtree-class": ",".join(sorted(self.child_classes)),
                          }
                else:
                self.update_qs(
                                {
                                    "rsp-subtree": "full",
                                    "rsp-subtree-class": ",".join(sorted(self.child_classes)),
                                }

                    
```

Reply:
I actually decided to do in the module to avoid touching this code, because sub-tree not supported for a class seems a corner case and which is saver to handle in the module instead of potentially changing behaviour.

I'm not sure I follow your conditional exactly but "rsp-subtree" in this specific case is a problem when the url is pointing to the root class. However the construct logic in this case seems already off, when root class provided, parent class not, and obj it self is also a constructs the url with the root class. We would then need to assure this conditional works on all the _construct_url functions in combination with the subclasses provided, and test this for all combinations. Not sure this is worth the effort for a corner case in my opinion.

If we decide to refactor I would say we would need to re-assess the entire url and filter construction mechanism, but  also there not sure this is worth it due to potential breaking changes.

Reply:
I thought the problem was when only "rsp-subtree" is defined? It's not set anywhere else in the code apart from construct_url()?
Why would all the instances of _construct_url() need a change?
If the issue is in the root class then the conditions can be changed.

However, if this is a corner case then it's not worth the effort I guess.

Reply:
I am not saying all instances of _construct_url() need a change I am saying we should test/assure that if we introduce a change like this we need to assure this does not impact current behaviour.

I think that the issue is indeed the root class in this case, but then the url constructed that is pointing to the root class. In other combinations like when you provide domain for instance it would actually request the url towards the domain ( parent class ), however when you provide the name ( obj ) with out domain it would query again on the root class in the url with filter. Thus in your conditional I am not sure that this error would be resolved, because when the subclass_3 ( name ) is provided, this class would not be in EXCLUDE_RSP_SUBTREE and thus you would insert rsp-subtree, however the url is still pointing to the root class. 

This is now only the case for the _construct_url_3() function, but how will this url construction behave for the other functions? This now is not required because it is the only case, but if we introduce this generic logic here I assume it should be working in every case. Also we would need to consider your nesting changes in this part from, https://github.com/CiscoDevNet/ansible-aci/pull/505/files for subclass_6.

Reply:
Yeah I think its better not to generalize this then.